### ISSUE TITLE
Optimisé le démarrage de l'application

### ISSUE THREAD

**Issue by detobel36** [RobinPetit/OpenWifiLocalizator#50]
Fichier avec le temps de chargement de chaque élément: [loadTime.txt](https://github.com/RobinPetit/OpenWifiLocalizator/files/761233/loadTime.txt)
          
## Ecran de chargement
On ajoute un écran de chargement au début de l'application

## Chargé les Wifi par la suite
Le but est de charger tous les noeuds et pas les wifi.  Pour connaitre la localisation on ferait un appel base de donnée qui retournerait toutes les nodes contenant au moins un des wifi capté par l'utilisateur.          
**Désavantage** on accède à la base de donnée à chaque fois qu'on fait une localisation (on est pas sur d'avoir tout chargé)

## Chargé les wifi au fur et à mesure.
Optimisation de la proposition précédente.
On charge les deux campus, on affiche donc l'application.          
On fait une requête pour charger les points susceptible d'être la bonne location (donc tout ceux ayant au moins un wifi en commun avec le wifi capté).           
*En même temps*, dans un **autre thread** on charge tous les noeuds/wifi.  Le temps que l'utilisateur se repère et pousse sur la recherche, on peut espérer avoir tout chargé...      
**Désavantage:** gérer la création "double" de noeud et être sur que l'utilisateur n'utilise pas l'app trop vite :disappointed_relieved:

**Comment by detobel36** [comment]
`I/Choreographer: Skipped 448 frames!  The application may be doing too much work on its main thread.`
L'idée des différent thread sera peut-être à envisager...

**Comment by denishoornaert** [comment]
Ne pas charger tout les wifi reste une bonne chose car même si pour le moment la différence n'est que de 1sec (ce qui est déja pas mal comme gain), la quantité de wifi présent en bdd va plus que surement "exploser".

**Comment by RobinPetit** [comment]
> Le but est de charger tous les noeuds et pas les wifi. Pour connaitre la localisation on ferait un appel base de donnée qui retournerait toutes les nodes contenant au moins un des wifi capté par l'utilisateur.
Désavantage on accède à la base de donnée à chaque fois qu'on fait une localisation (on est pas sur d'avoir tout chargé)

Et si on chargeait d'abord uniquement les wifis de la première localisation, puis qu'une fois l'utilisateur localisé, on lance (dans un autre thread ?) le chargement des autres wifis ?

Ça permettrait à l'application de *démarrer vite* dans le sens où il n'y aurait pas 50 minutes de chargement au démarrage, mais il ne faudrait pas refaire des chargements à chaque localisation.

Vous en pensez quoi ?

**Comment by RobinPetit** [comment]
Je pensais à un truc hier en prenant mon train : on ne pourrait pas charger le graphe avec tous les nœuds sans récupérer les wifis, et les récupérer uniquement au fur et à mesure qu'on localise ?

Genre on capte n wifis w_1 à w_n à un point quelque part dans le batiment H, donc pas besoin de charger tous les wifis de la plaine ou les wifis lointains. Du coup, au prochain scan, on re-récupère uniquement les wifis nécessaires, etc.

Vous en pensez quoi ? Je pense que ça va pouvoir méchemment accélérer le démarrage.

**Comment by detobel36** [comment]
C'est déjà ce qu'on fait 😝

**Comment by RobinPetit** [comment]
Il me semblait que l'on chargeait uniquement les wifis nécessaires au démarrage, mais qu'après on les chargeait tous tout de même

**Comment by detobel36** [comment]
C'est le but au final mais que j'ai pas encore implémenter...           
C'est à dire que maintenant on appel quoi qu'il arrive la BDD pour charger tous les wifi des nodes qui ont au moins un wifi en commun avec les wifis captés.             
Ca m'embête pas mal qu'on appel à chaque fois la bdd 😕  D'où le fait que je proposais qu'on ai une tâche de fond à la fin qui charge tous les wifis...

**Comment by RobinPetit** [comment]
Mais en même temps, les wifis, on en a uniquement besoin lorsque l'on localise, du coup, je ne pense pas qu'il soit nécessaire de tout garder dans la mémoire du téléphone. En plus pour un phone avec peu de RAM, ça va surcharger inutilement à mon sens

**Comment by detobel36** [comment]
Ouep, du coup je peux modifier légèrement mon code (que j'avais préparé pour ce chargement en "background") pour que tout soit propre.            
Et du coup on pourra close cette issue 😄

**Comment by RobinPetit** [comment]
Yep, sinon je m'étais dit que j'allais également voir quels types étaient utilisés où, voir s'il n'y avait pas moyen d'un peu alléger également à ce niveau là : ne pas utiliser systématiquement des ArrayList si des [] suffisent (sans tomber dans des optimisation incompréhensibles non plus...)

**Comment by RobinPetit** [comment]
Je pense qu'ici il va vraiment falloir faire quelque chose... Pour aujourd'hui, je ferai des benchmarks pour voir ce qui prend du temps, et où, puis essayer de corriger déjà ce que je peux
